File: 0078-subsets/0078-subsets.java
class Solution {
    
    List<List<Integer>> result;
    int k;
    
    public List<List<Integer>> subsets(int[] nums) {
        result = new ArrayList<>();
        
        for(k = 0; k <= nums.length; k++)
        {
            dfs(0, new ArrayList<Integer>(), nums);
        }
        
        return result;
    }
    
    
    private void dfs(int start, List<Integer> curt, int[] nums)
    {
        if(curt.size() == k)
        {
            result.add(new ArrayList<Integer>(curt));
            return;
        }
        
        for(int i = start; i < nums.length; i++)
        {
            curt.add(nums[i]);
            dfs(i + 1, curt, nums);
            curt.remove(curt.size() - 1);
        }
        
    }
}

File: 0078-subsets/0078-subsets.cpp
class Solution {
public:
    void dfs(int start, vector<int>& nums, vector<int>& result, vector<vector<int>>& results, int k)
    {
        if(result.size() == k)
        {
            results.push_back(result);
            return;
        }
        
        for(int i = start; i < nums.size(); i++)
        {
            result.push_back(nums[i]);
            dfs(i + 1, nums, result, results, k);
            result.pop_back();
        }
    }
    
    vector<vector<int>> subsets(vector<int>& nums)
    {
        vector<vector<int>> results;
        vector<int> result;
        
        for(int k = 0; k <= nums.size(); k++)
            dfs(0, nums, result, results, k);
        return results;
    }
};